### DISCUSSION TITLE
Ditch the installer for a simple zip and move the build to gradle?

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#563] [Development Practices]
@opendcs/opendcs-core-devs @opendcs/opendcs-triage @opendcs/opendcs-dcpmon-devs 

Okay, sorry Karl keeps bringing this up and frankly I don't disagree but it's not just on the two of us.

What does everyone think about ditching the installer? 
Justification:
- Karl already fixed the windows batch files to not need the full path and can do so to the bash scripts
- The integration tests don't split out the implementations so we know they aren't interfering with each other
- 7.0.13, when final, will have proper support for DCSTOOL_USERDIR in all environments.
- It'll make things easier

Justification for gradle:
- Better test result output
- Sane defaults for file locations/multi project (we're already using them but I have to manually manage the ant build design)
- Additional tools for various static and dependency analysis and reporting.
- Less manual setup to start development

The main reason we stuck with ant was that Mike M was having issues with Eclipse not dealing with gradle well.
1) Mike M removed himself as a core developer
2) Eclipse handles gradle better (which is good, because I want to use the WindowsBuilder to cleanup existing and create some new GUIs)

**Comment by ktarbet** [comment]
Here are some potential concerns on Zip file.   (note: I'll **all** in favor of zip file)

* The unzip/tar would install both CWMS and HDB and XML  [ this created conflicts in older versions ]
* Windows users may like an installer.
* A large warning can occur in Windows 11 when running apps unzipped with windows extract.  

Potential resolution of above concerns:

* I've heard the HDB/CWMS conflicts are no longer an issue.
* We can make a windows only installer if demand warrants; there are many tools available for Windows installers.
* Use 7-zip to install;  sign the application so Windows Defender will trust the app.


@MikeNeilson  Please remind us why we are lumping zip file install, and gradle in to a single discussion.  Is it because gradle is going to make the creating a zip installer easy?

**Reply by MikeNeilson** [reply]
Mostly becomes my main concern with moving to gradle is making sure the installer doesn't get screwed up, no installer, less concern.

**Reply by DanielTOsborne** [reply]
When you say 7-zip, do you mean making a self-extracting exe?

The flag windows uses to mark an internet download is in an alternate data stream. So if you want to keep it simple, you could have a batch script that strips that alternate data stream from the files.
Extract zip, run batch script, done.

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/unblock-file?view=powershell-7.4

**Comment by DanielTOsborne** [comment]
> Justification for gradle:
> 
> * Better test result output
> * Sane defaults for file locations/multi project (we're already using them but I have to manually manage the ant build design)
> * Additional tools for various static and dependency analysis and reporting.
> * Less manual setup to start development
> 

I'll preface this saying I'm biased _against_ gradle for multiple reasons.

I'm not sure what you mean by "better test result output." I'm trying to parse test output in hec-cwms-data-access, which also uses gradle, and the results are pretty unreadable, particularly if you need to compare between two test runs.
Maybe the HTML report that's produced is nice, but trying to parse tests through any other method, like diff, sed, awk, etc, is nearly impossible. I can't even get a simple list of all the tests that failed in a run. I don't want any other details, just the list of tests.


Regarding gradle itself (applies to both groovy and kotlin):
I find the syntax terse and difficult to read/understand what's going on. It relies on a lot of assumed magic (from my perspective).

It doesn't use a descriptive/markup language, but a compiled/interpreted and executed language (as best I understand it anyway).
Practically speaking, it's probably not a problem, but it just is not something I agree with personally in a build system.

Compare this with something like maven (which I personally prefer, but am not pushing as an alternative in this post), which is much more explicit about the definitions. Yes, it's more verbose, particularly since it's XML. However when you import plugins, then configure/use them, it's crystal clear what plugin is being used for that section of the build. I don't have to look it up, or have assumed knowledge about the language and system to know how to read the build file.

Where this has been a problem for me before is when I try to run a build that's broken. For example, if the javadoc tasks fail. There was no explicit lines in the gradle files for javadocs, so I had to look up how to disable the default behavior of building javadocs, just so I could test the project, since the tests wouldn't run if the javadoc build failed.

Gradle seems fine as long as you're on the "happy path." Stray from that, even a little, and it's a nightmare.

**Reply by MikeNeilson** [reply]
`Gradle seems fine as long as you're on the "happy path." Stray from that, even a little, and it's a nightmare`

My experience has shown that to be far more true of Maven than Gradle. Which is way most of the HEC software uses it.

Personally I also find the Groovy easier to read than the Maven XML which is purely descriptive. The Ant XML isn't too bad but it is a *lot* to maintain. Yeah, Gradle and Maven *both* hide a lot, but that's kind of the point. if you want to see what they're hiding, go look in detail at the opendcs build.xml and build-macros.xml and common.xml.

But on the tests, yes our current build does give that simple list, but that's it. Go try limiting to a single test (you actually can do that now, but there's quite a few options in Junit for which tests to run which can be really helpful in the integration tests to save a lot of time.)

The cwms-data-api build isn't a good example, we really need to cleanup that output, and we have a bit, but I like that it puts the error inline and I don't have to hunt for it. However, if you really only want that, pretty sure a user gradle.properties or init would let you configure that.

**Reply by MikeNeilson** [reply]
That all said, my main concern is IDE integration. The current build requires the absolute latest Apache Ant to use and thus requires some manual interaction by the user. Both Maven and Gradle include a wrapper, plus the various IDEs keep up better. 
So I'm not dead set on gradle, though I'm far more comfortable with it, but I do think moving on from Ant will help bringing people on.
Perhaps a poll of all of us to get a sense of preference.

Though perhaps as a selling point, the 8.0 branch is already gradle, so the bulk of the work has been done.

**Reply by adamkorynta** [reply]
After migrating way too many legacy projects from ant to maven, I find gradle far easier to use for anything other than "compile source code and build a jar". It was a hard sell for me at first as gradle seemed like a lot of magic and I understood maven's build cycle much better, but once HEC moved to gradle, writing application installer bundle scripts became much simpler and allowed us to delete a bunch of custom written code for maven. 

The other sell for me is that IntelliJ's maven integration is not quite as tight as its gradle integration (it doesn't bind the default build/test runs to maven tasks). It seems Eclipse, VSCode and NetBeans have the same level of integration for either build tool. Lastly, debugging an applications through the IDE in maven required separate remote debugger configuration, meaning I'd have to run the application, then attach a remote debugger. I don't know if that's changed in the past two years or so, but that's a pretty annoying barrier to entry compared to just clicking the debug button (https://spin.atomicobject.com/maven-debugging-intellij/).

**Reply by DanielTOsborne** [reply]
I can't speak to other IDEs, as the only other one I've really used to some extent for Java is Eclipse, and I haven't used that for a long time. However VSCode's Gradle integration seems inferior to Maven's. With Maven build, everything is integrated into the regular VSCode workflow. With gradle, everything is inside the gradle extension, separate from the main IDE.

One thing I've had a problem with is that if I run a gradle build from the terminal, VSCode doesn't see it as a built project. Meaning I don't get dependencies pulled in for intellisense. Other cases I've had problems with is the build failing in the IDE, but working fine in the command-line. Perhaps something is configured wrong, but that's my point: the integration should handle the configuration automatically. I shouldn't need to independently configure the build and the IDE, which pulls from the build anyway.
Perhaps there's a better extension than Microsoft's one though.

> Lastly, debugging an applications through the IDE in maven required separate remote debugger configuration, meaning I'd have to run the application, then attach a remote debugger. I don't know if that's changed in the past two years or so, but that's a pretty annoying barrier to entry compared to just clicking the debug button (https://spin.atomicobject.com/maven-debugging-intellij/).

Funny enough, that's the problem I have with Gradle in VSCode, due to the above issues. I can't debug anything that uses Gradle. If I'm lucky I can run tests in the IDE.

**Comment by ktarbet** [comment]
We have two discussions going here.   I haven't heard anything against using a zip file instead of having an installer.

I was thinking just a zip file - NOT self extracting.   

I'll create a Issue for the zip file and retirement of izpack.

**Comment by MikeNeilson** [comment]
So I've only heard one dissenting opinion on gradle, and others were in favor. I'm going to go ahead and make a PR with the change and we can discuss further from there. 

For this PR I'll keep the installer as gradle can call ant tasks.

**Comment by kakashianh** [comment]
Can i debug kotlin (vscode) in spring boot maven java (i only use kotlin build controller)


https://stackoverflow.com/questions/78755089/debug-kotlin-in-spring-boot-maven-java

**Reply by MikeNeilson** [reply]
I would assume so, however we use neither Kotlin, spring boot, or maven, in this project and we are thus not the people to ask if and how.
